src/Features/Core/Portable/Formatting/FormattingDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,101 @@
+
+        public FormattingDiagnosticAnalyzer()
+            : base(
+                IDEDiagnosticIds.FormattingDiagnosticId,
+                new LocalizableResourceString(nameof(FeaturesResources.Formatting_analyzer_title), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+                new LocalizableResourceString(nameof(FeaturesResources.Formatting_analyzer_message), FeaturesResources.ResourceManager, typeof(FeaturesResources)))
+        {
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SyntaxAnalysis;
+
+        public override bool OpenFileOnly(Workspace workspace)
+            => false;
+
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterSyntaxTreeAction(AnalyzeSyntaxTree);
+
+        private void AnalyzeSyntaxTree(SyntaxTreeAnalysisContext context)
+        {
+            if (!(context.Options is WorkspaceAnalyzerOptions workspaceAnalyzerOptions))
+            {
+                return;
+            }
+
+            var options = context.Options.GetDocumentOptionSetAsync(context.Tree, context.CancellationToken).GetAwaiter().GetResult();
+            if (options == null)
+            {
+                return;
+            }
+
+            var workspace = workspaceAnalyzerOptions.Services.Workspace;
+            var formattingChanges = Formatter.GetFormattedTextChanges(context.Tree.GetRoot(context.CancellationToken), workspace, options, context.CancellationToken);
+            foreach (var formattingChange in formattingChanges)
+            {
+                var change = formattingChange;
+                if (change.NewText.Length > 0 && !change.Span.IsEmpty)
+                {
+                    var oldText = context.Tree.GetText(context.CancellationToken);
+
+                    // Handle cases where the change is a substring removal from the beginning
+                    var offset = change.Span.Length - change.NewText.Length;
+                    if (offset >= 0 && oldText.GetSubText(new TextSpan(change.Span.Start + offset, change.NewText.Length)).ContentEquals(SourceText.From(change.NewText)))
+                    {
+                        change = new TextChange(new TextSpan(change.Span.Start, offset), "");
+                    }
+                    else
+                    {
+                        // Handle cases where the change is a substring removal from the end
+                        if (change.NewText.Length < change.Span.Length
+                            && oldText.GetSubText(new TextSpan(change.Span.Start, change.NewText.Length)).ContentEquals(SourceText.From(change.NewText)))
+                        {
+                            change = new TextChange(new TextSpan(change.Span.Start + change.NewText.Length, change.Span.Length - change.NewText.Length), "");
+                        }
```

Comment:
this is basically offset?

Reply:
Covered by https://github.com/dotnet/roslyn/pull/28873#discussion_r208386096